astropy/io/fits/file.py

```diff
@@ -109,15 +109,19 @@ def __init__(self, fileobj=None, mode=None, memmap=None, overwrite=False,
 
         if mode is None:
             if _is_random_access_file_backed(fileobj):
-                fmode = fileobj_mode(fileobj)
-                # If the mode is unsupported just leave it as None; we'll
-                # catch this case below
-                mode = FILE_MODES.get(fmode)
+                mode = fileobj_mode(fileobj)
             else:
                 mode = 'readonly'  # The default
 
         if mode not in IO_FITS_MODES:
-            raise ValueError("Mode '{}' not recognized".format(mode))
+            if mode in ['r', 'w', 'a']:
```

Comment:
You are right about the `"*t"`options: they do need to be handled (I had forgotten that was allowed actually). This was really just an attempt to refactor the existing logic to be a little bit less redundant, and personally I think it's better with the more explicit error message.